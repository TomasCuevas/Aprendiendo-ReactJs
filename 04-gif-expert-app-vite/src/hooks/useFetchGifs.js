import { useEffect, useState } from 'react';

/**
 * @helpers
 */
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {
  const [ gifs, setGifs ] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  
  const getCurrentGifs = async() => {
    const gifs = await getGifs(category);
    setGifs(gifs);
    setIsLoading(false);
  }
  
  useEffect(() => {
    getCurrentGifs();
  }, [])

  return { gifs, isLoading };
}