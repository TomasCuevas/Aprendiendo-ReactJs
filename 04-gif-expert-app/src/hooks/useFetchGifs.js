import { useEffect, useState } from "react"
import { getFetchGifs } from "../helpers/getFetchGifs";

export const useFetchGifs = ( category ) => {

	const [ state, setState ] = useState({
		data: [],
		loading: true
	});

	useEffect( () => {

		getFetchGifs( category )
			.then( ( imgs ) => {
				setTimeout( () => {

					setState({
						data: imgs,
						loading: false
					})
					
				}, 3000)
			});
	}, [ category ]);

	return state;
}