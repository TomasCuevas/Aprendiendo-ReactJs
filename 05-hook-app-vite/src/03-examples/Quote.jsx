import { useLayoutEffect, useRef, useState } from 'react';

export const Quote = ({ author, quote }) => {
  const [boxSizing, setBoxSizing] = useState({ width: 0, height: 0 });
  const pRef = useRef()

  useLayoutEffect(() => {
    const { width, height } = pRef.current.getBoundingClientRect();
    setBoxSizing({ width, height });
  }, [quote])
  
  return (
    <>
      <blockquote 
        className="blockquote text-end"
        style={{ display: 'flex' }}
      >
        <p ref={pRef} className="mb-2">
          { quote }
        </p>
        <footer className="blockquote-footer">
          { author }
        </footer>
      </blockquote>

      <code>
        { JSON.stringify(boxSizing) }
      </code>
    </>
  )
}