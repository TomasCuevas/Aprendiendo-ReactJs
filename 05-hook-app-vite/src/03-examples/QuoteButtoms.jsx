export const QuoteButtoms = ({ counter, decrement, increment }) => {

  return (
    <>
      <button 
        className="btn btn-primary"
        onClick={() => decrement()}
        disabled={counter === 1 && true}
      >
        Previous Quote
      </button>
      <button 
        className="btn btn-primary"
        onClick={() => increment()}
      >
        Next Quote
      </button>
    </>
  )
}