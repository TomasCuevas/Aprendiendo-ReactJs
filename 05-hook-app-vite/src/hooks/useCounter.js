import { useState } from 'react';

export const useCounter = ( initialState = 0 ) => {
  const [ counter, setCounter ] = useState(initialState);

  return {
    counter
  }
}