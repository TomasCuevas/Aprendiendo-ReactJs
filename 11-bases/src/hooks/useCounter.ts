import { useState, useRef, useEffect } from 'react';
import gsap from 'gsap';

interface Props {
  MAXIMUN_COUNT?: number;
}

export const useCounter = ({ MAXIMUN_COUNT = 10 }: Props) => {
  const [counter, setCounter] = useState(5);
  const counterElement = useRef<HTMLHeadingElement>(null);

  const onIncrement = () => {
    setCounter((prev) => Math.min(prev + 1, MAXIMUN_COUNT));
  };

  useEffect(() => {
    if (counter < 10) return;

    console.log('%cSe llego al valor maximo', 'color: red; background-color: black');

    const timeline = gsap.timeline();

    timeline
      .to(counterElement.current, { y: -10, duration: 0.2 })
      .to(counterElement.current, { y: 10, duration: 1, ease: 'bounce.out' });
  }, [counter]);

  return {
    // properties
    counter,
    counterElement,

    // methods
    onIncrement,
  };
};
