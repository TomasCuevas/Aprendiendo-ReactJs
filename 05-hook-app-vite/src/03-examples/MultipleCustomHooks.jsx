import { useCounter, useFetch } from '../hooks';
import { LoadingQuote, Quote, QuoteButtoms } from './';

export const MultipleCustomHooks = () => {
  const { counter, increment, decrement } = useCounter(1);
  const { data, isLoading, hasError } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

  const { quote, author } = !!data && data[0];

  return (
    <>
      <h1>Breaking Bad Quotes</h1>
      <hr />

      { (isLoading) && <LoadingQuote />}

      { (!isLoading) && (
        <>
          <Quote 
            quote={quote} 
            author={author} 
          />
          <QuoteButtoms 
            counter={counter} 
            increment={increment} 
            decrement={decrement} 
          />
        </>
      )}
    </>
  )
}