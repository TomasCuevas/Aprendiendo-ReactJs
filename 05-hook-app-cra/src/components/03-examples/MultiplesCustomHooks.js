import React from 'react'
import { useCounter } from '../../hooks/useCounter';

import { useFetch } from '../../hooks/useFetch';
import './effects.css';

export const MultiplesCustomHooks = () => {
    const { counter, increment } = useCounter(1);
    const { data, loading }      = useFetch( `https://www.breakingbadapi.com/api/quotes/${ counter }` );

    const { author, quote } = !!data && data[0];
    
    return (
        <>  
            <h1 className='title'>Breaking Bad Quotes</h1>
            <hr/>

            {
                ( loading )
                ?
                    (
                        <div className='loading__container'>
                            <span>Loading...</span>
                        </div>
                    )
                : 
                    (
                        <blockquote className='blockquote__container'>
                            <p className='quote'>{ quote }</p>
                            <footer className='author'> - { author }</footer>
                        </blockquote>
                    )
            }
            <button 
                className='next__quoter'
                onClick={ increment }
            >
                Siguiente Frase
            </button>
        </>
    )
}
