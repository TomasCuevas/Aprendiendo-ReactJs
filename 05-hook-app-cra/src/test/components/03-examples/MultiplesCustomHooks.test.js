import React from 'react';
import { shallow } from "enzyme";

import { MultiplesCustomHooks } from "../../../components/03-examples/MultiplesCustomHooks";
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultiplesCustomHooks />', () => {

  test('debe de mostrarse correctamente', () => {    

    useCounter.mockReturnValue({
      counter: 1,
      increment: () => {}
    });
    
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })
    
    const wrapper = shallow(<MultiplesCustomHooks />);
    expect( wrapper ).toMatchSnapshot();
  });

  test('debe de mostrar la informacion', () => {

    useCounter.mockReturnValue({
      counter: 1,
      increment: () => {}
    });
    
    useFetch.mockReturnValue({
      data: [{
        author: 'Anselmo',
        quote: 'Hola Mundo'
      }],
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultiplesCustomHooks />);

    expect(wrapper.find('.loading__container').exists()).toBe(false);
    expect(wrapper.find('.quote').text()).toBe('Hola Mundo');
    expect(wrapper.find('.author').text()).toBe(' - Anselmo');
  });
})