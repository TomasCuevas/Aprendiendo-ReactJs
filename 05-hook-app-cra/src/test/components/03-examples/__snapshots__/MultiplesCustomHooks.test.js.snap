// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultiplesCustomHooks /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h1
    className="title"
  >
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="loading__container"
  >
    <span>
      Loading...
    </span>
  </div>
  <button
    className="next__quoter"
    onClick={[Function]}
  >
    Siguiente Frase
  </button>
</Fragment>
`;
